### DISCUSSION TITLE
Optional Inputs and outputs

### DISCUSSION THREAD

**Discussion by mattjoneswo** [terjeio/ioSender#109] [Ideas]
Loving the grblhal and iosender.   How difficult would it be to add the “$S” command single step/block mode as an input for physical button.  I assume would correlate with the start & feedhold chain.  

Also, is it possible to send status signals to an output on the breakout board?   I am using Phil’s teenage 4.1 breakout board with a few available outputs.   Would be nice to send signal to an LED either on the button or a light tree for status such as running/paused (feed hold)/ single block on, etc.

**Comment by phil-barrett** [comment]
Extra LEDs are very easy to add via the I2C header.  I have one design that brings out 8 LEDs + 8 GPIOs (LEDs + resistors) via an I2C daughter card.  The LEDs can physically be anywhere (well, within say 2 M of wire).  Then M6[2-5] commands could be used to turn the lights on or off (would need a bit of SW work). It is shown below.  Works well with the hardware.  There are power limitations to this approach so it needs an external power source if more than 8 LEDs would be on simultaneously or something like 12V is needed for an illuminated switch.  Easy enough to do.  Food for thought.

![image](https://user-images.githubusercontent.com/16570958/120016605-a7bcc580-bf99-11eb-8832-e54a32c11c77.png)

**Comment by terjeio** [comment]
> How difficult would it be to add the “$S” command single step/block mode as an input for physical button. I assume would correlate with the start & feedhold chain.

A check box is IMO more appropriate, like _Flood_ and _Mist_. Not very difficult and _Block delete_ should be added to the same control. I have been thinking about adding this for a long time (since I added _Block delete_ support to the grblHAL core).

> Also, is it possible to send status signals to an output on the breakout board?

Yes, but that should be done on the controller side with a plugin. A plugin can subscribe to events in the core and control LEDs (or other outputs) from those. This should done by claiming ioport pins and calling functions via the HAL API to set/clear them. There is no need to use M-codes for that. BTW the M6[2-5] commands uses the very same same API...

Writing such a plugin could be used as a case study for how add functionality to grblHAL? I would like somebody else than me (you?) to write such a plugin. I can provide help - it is not going to be much code. Nice if it could end up as a blog post for Phil?

**Comment by terjeio** [comment]
FYI all optional control signals can now be bound to aux inputs with [compile time options](https://github.com/grblHAL/core/blob/master/changelog.md#20231229). I will add a config page to the Web Builder later.
